### PR TITLE
DOC Rework `GapEncoder` example

### PR DESCRIPTION
**Pull request by LilianBoulard**
Extracted from #546.

### CODE REVIEW

File: examples/02_investigating_dirty_categories.py [hunk 1]
```diff
@@ -1,188 +1,60 @@
...
-values.insert(0, "current_annual_salary", employee_salaries.y)
-
-###############################################################################
-# .. _example_similarity_encoder:
-#
-# String similarity between entries
-# ---------------------------------
+# The |GapEncoder| is a better encoder than the |SimilarityEncoder| in the
+# sense that it is more scalable and interpretable,
+# which we will demonstrate now.
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:22]:
```suggestion
# The |GapEncoder| is more scalable and interpretable than the |SimilarityEncoder|,
# which we will demonstrate now.
```

File: examples/02_investigating_dirty_categories.py [hunk 2]
```diff
@@ -1,188 +1,60 @@
...
 #
-# That's where our encoders get into play.
-# In order to robustly embed dirty semantic data, the |SE|
-# creates a similarity matrix based on an n-gram representation of the data.
-
-sorted_values = values["employee_position_title"].sort_values().unique()
-
-from skrub import SimilarityEncoder
+# First, let's retrieve the dataset:
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:24]:
```suggestion
# First, let's retrieve the `employee salaries dataset <https://www.openml.org/d/42125>`_:
```

File: examples/02_investigating_dirty_categories.py [hunk 3]
```diff
@@ -1,188 +1,60 @@
...
-print(two_dim_data.shape)
-print(sorted_values.shape)
+dataset.description
 
 ###############################################################################
-# We first quickly fit a KNN so that the plots does not get too busy:
-
-import numpy as np
-from sklearn.neighbors import NearestNeighbors
+# And carry out some basic preprocessing
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:36]:
```suggestion
# And carry out some basic preprocessing:
```

File: examples/02_investigating_dirty_categories.py [hunk 4]
```diff
@@ -1,188 +1,60 @@
...
-#
+X
 
 ###############################################################################
-# The |Gap| is a better encoder than the
-# |SE| in the sense that it is more scalable and
-# interpretable, which we will present now.
-#
-# First, let's retrieve the dirty column to encode:
+# Let's also get the dirty column we want to encode
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:48]:
```suggestion
# Let's look at the job titles, the dirty column we want to encode:
```

File: examples/02_investigating_dirty_categories.py [hunk 5]
```diff
@@ -1,188 +1,60 @@
...
 
 ###############################################################################
-# .. _example_gap_encoder:
-#
 # Encoding dirty job titles
-# .........................
+# -------------------------
 #
-# Then, we'll create an instance of the |Gap| with 10 components:
+# Then, we'll create an instance of the |GapEncoder| with 10 components:
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:57]:
```suggestion
# Then, we'll create an instance of the |GapEncoder| with 10 components. This means that the encoder will attempt to extract 10 latent topics from the input data:
```

File: examples/02_investigating_dirty_categories.py [hunk 6]
```diff
@@ -219,6 +90,8 @@
 #
 # This enables us to understand the encoding of different samples
```
Comment (jovan-stojanovic) [inline examples/02_investigating_dirty_categories.py:91]:
```suggestion
# This enables us to understand the encoding of different samples.
```

**Review by jovan-stojanovic — CHANGES_REQUESTED** [review]
Thanks @LilianBoulard, this looks very good! A few comments.

File: examples/02_feature_interpretation_with_gapencoder.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+"""
+.. _example_interpreting_gap_encoder:
+
+==========================================
+Feature interpretation with the GapEncoder
+==========================================
+
+In this notebook, we will explore the output and inner workings of the
+|GapEncoder|, one of the encoders provided by skrub.
```
Comment (GaelVaroquaux) [inline examples/02_feature_interpretation_with_gapencoder.py:9]:
```suggestion
|GapEncoder|, one of the categorical encoders provided by skrub.
```

File: examples/02_feature_interpretation_with_gapencoder.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+.. |GapEncoder| replace::
+     :class:`~skrub.GapEncoder`
+
+.. |SimilarityEncoder| replace::
+     :class:`~skrub.SimilarityEncoder`
+"""
+
+###############################################################################
+# The |GapEncoder| is more scalable and interpretable than the
+# |SimilarityEncoder|, which we will demonstrate now.
```
Comment (GaelVaroquaux) [inline examples/02_feature_interpretation_with_gapencoder.py:21]: Let's not position the GapEncoder with regards to the SimilarityEncoder, as we are no longer putting emphasis on the latter
```suggestion
# The |GapEncoder| is scalable and interpretable in terms of
# finding latent categories, as we will show.
```

File: examples/02_feature_interpretation_with_gapencoder.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+X_dirty = X[["employee_position_title"]]
+
+###############################################################################
+# Let's have a look at a sample of the job titles:
+
+X_dirty.str.contains("Fire").sample(10, random_state=0).head(10)
+
+###############################################################################
+# Then, we'll create an instance of the |GapEncoder| with 10 components.
```
Comment (GaelVaroquaux) [inline examples/02_feature_interpretation_with_gapencoder.py:61]:
```suggestion
# Then, we create an instance of the |GapEncoder| with 10 components.
```

File: examples/02_feature_interpretation_with_gapencoder.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+# This means that the encoder will attempt to extract 10 latent topics
+# from the input data:
+
+from skrub import GapEncoder
+
+enc = GapEncoder(n_components=10, random_state=1)
+
+###############################################################################
+# Finally, we'll fit the model on the dirty categorical data and transform it
```
Comment (GaelVaroquaux) [inline examples/02_feature_interpretation_with_gapencoder.py:70]:
```suggestion
# Finally, we fit the model on the dirty categorical data and transform it
```

**Review by GaelVaroquaux — COMMENTED** [review]
LGTM, but I think that we also need to change the index because there is a link to the example which has been renamed, no?

**Comment by jovan-stojanovic** [comment]
You should adapt the example to the fact that #581 was merged

**Comment by jovan-stojanovic** [comment]
Thanks @LilianBoulard! I will take over this PR so we move faster.

File: examples/02_feature_interpretation_with_gapencoder.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+# on a set of latent topics estimated from the data. The latent topics
+# are built by capturing combinations of substrings that frequently
+# co-occur, and encoded vectors correspond to their activations.
+# To interpret these latent topics, we select for each of them a few labels
+# from the input data with the highest activations.
+# In the example below we select 3 labels to summarize each topic.
+
+topic_labels = enc.get_feature_names_out(n_labels=3)
+for k, labels in enumerate(topic_labels):
+    print(f"Topic n°{k}: {labels}")
```
Comment (Vincent-Maladiere) [inline examples/02_feature_interpretation_with_gapencoder.py:78]: This part (and the subsequent visualization) is great! A real wow effect 👍

**Review by Vincent-Maladiere — COMMENTED** [review]
Great concision effort again (-100!). My main remark regards using "dirty" either for variable names or explainers. This can be easily misinterpreted, and I'd rather have "high cardinality" instead or "noisy" if you prefer. WDYT?

**Comment by jovan-stojanovic** [comment]
> My main remark regards using "dirty" for variable names and explainers. This can be easily misinterpreted, and I'd rather have "high cardinality, "high entropy" or "noisy" instead. WDYT?

You have a point, looking back at the example I realize that we introduce "dirty" without explaining what it means. I will correct this.

This term "dirty" is something that [was used in the literature](https://inria.hal.science/hal-01806175) and in dirty-cat to describe data whose source of variation can be anything, meaningful and meaningful as opposed to "noisy" for instance, which has different connotation (often synonym to meaningless) in some fields (e.g. econometrics, signal processing).

As for alternatives, I like "high cardinality" ([also used in literature](https://inria.hal.science/hal-02171256v4)), "high entropy" may be less known among our users.

**Comment by jovan-stojanovic** [comment]
WDYT of this version? @Vincent-Maladiere

**Comment by Vincent-Maladiere** [comment]
> This term "dirty" is something that [was used in the literature](https://inria.hal.science/hal-01806175) and in dirty-cat to describe data whose source of variation can be anything

Yes, we used it in a paper, but it doesn't mean it conveys the right message 😅. External users might be surprised by this wording, like Cailean Osborne —he was in high spirits when he found out we call those "dirty"!

> WDYT of this version? @Vincent-Maladiere

Well, I still prefer using "high cardinality" if you like. Let's bring this up during the next meeting.

**Review by Vincent-Maladiere — APPROVED** [review]
LGTM!

**Pull request merged**